ext/intl/dateformat/datepatterngenerator_methods.cpp

```diff
@@ -0,0 +1,145 @@
+#include <assert.h>
+}
+
+using icu::DateTimePatternGenerator;
+using icu::Locale;
+using icu::StringPiece;
+
+static zend_result dtpg_ctor(INTERNAL_FUNCTION_PARAMETERS)
+{
+	zval		*object;
+	char		*locale_str;
+	size_t		locale_len	= 0;
+	Locale		locale;
+	IntlDatePatternGenerator_object* dtpgo;
+
+	intl_error_reset(NULL);
+	object = return_value;
+
+	ZEND_PARSE_PARAMETERS_START(0, 1)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_STRING_OR_NULL(locale_str, locale_len)
+	ZEND_PARSE_PARAMETERS_END_EX(return FAILURE);
+
+	DTPATTERNGEN_METHOD_FETCH_OBJECT_NO_CHECK;
+
+	if (dtpgo->dtpg != NULL) {
+		intl_errors_set(DTPATTERNGEN_ERROR_P(dtpgo), U_ILLEGAL_ARGUMENT_ERROR, "Cannot call constructor twice", 0);
+		return FAILURE;
+	}
+
+	INTL_CHECK_LOCALE_LEN_OR_FAILURE(locale_len);
+	if (locale_len == 0) {
+		locale_str = (char *) intl_locale_get_default();
+	}
+	locale = Locale::createFromName(locale_str);
+
+	dtpgo->dtpg = DateTimePatternGenerator::createInstance(
+		locale,
+		DTPATTERNGEN_ERROR_CODE(dtpgo));
+
+	if (U_FAILURE(DTPATTERNGEN_ERROR_CODE(dtpgo))) {
+		intl_error_set(NULL, DTPATTERNGEN_ERROR_CODE(dtpgo),
+				"Error creating DateTimePatternGenerator",
+				0);
+		goto error;
+	}
+
+error:
+	return U_FAILURE(intl_error_get_code(NULL)) ? FAILURE : SUCCESS;
+}
+
+U_CFUNC PHP_METHOD( IntlDatePatternGenerator, create )
+{
+    object_init_ex( return_value, IntlDatePatternGenerator_ce_ptr );
+    if (dtpg_ctor(INTERNAL_FUNCTION_PARAM_PASSTHRU) == FAILURE) {
+		zval_ptr_dtor(return_value);
+		RETURN_NULL();
+	}
+}
+
+U_CFUNC PHP_METHOD( IntlDatePatternGenerator, __construct )
+{
+	zend_error_handling error_handling;
+
+	zend_replace_error_handling(EH_THROW, IntlException_ce_ptr, &error_handling);
+	/* return_value param is being changed, therefore we will always return
+	 * NULL here */
+	return_value = ZEND_THIS;
+	if (dtpg_ctor(INTERNAL_FUNCTION_PARAM_PASSTHRU) == FAILURE) {
+		if (!EG(exception)) {
+			zend_string *err = intl_error_get_message(NULL);
+			zend_throw_exception(IntlException_ce_ptr, ZSTR_VAL(err), intl_error_get_code(NULL));
+			zend_string_release_ex(err, 0);
+		}
+	}
+	zend_restore_error_handling(&error_handling);
+}
+
+
+U_CFUNC PHP_METHOD( IntlDatePatternGenerator, getBestPattern )
+{
+	char			*skeleton_str	= NULL;
+	size_t			skeleton_len;
+	UnicodeString	skeleton_uncleaned;
+	UnicodeString	skeleton;
+
+	DTPATTERNGEN_METHOD_INIT_VARS;
+
+	/* Parse parameters. */
+	if( zend_parse_method_parameters( ZEND_NUM_ARGS(), getThis(), "Os",
+		&object, IntlDatePatternGenerator_ce_ptr,  &skeleton_str, &skeleton_len ) == FAILURE )
+	{
+		RETURN_THROWS();
+	}
+
+	DTPATTERNGEN_METHOD_FETCH_OBJECT;
+
+	intl_stringFromChar(skeleton_uncleaned, skeleton_str, skeleton_len, DTPATTERNGEN_ERROR_CODE_P(dtpgo));
+
+	INTL_METHOD_CHECK_STATUS(dtpgo, "Skeleton is not a valid UTF-8 string");
+
+	skeleton = dtpgo->dtpg->getSkeleton(skeleton_uncleaned, DTPATTERNGEN_ERROR_CODE(dtpgo));
+
+	INTL_METHOD_CHECK_STATUS(dtpgo, "Error getting cleaned skeleton");
+
+	{
```

Comment:
Code block unnecessary. As usual, you can directly combine declaration and initialization.